File: backend/app/api/contribute.py
"""The contribution panel: visitors fill in missing statements.

For historical scans, place and year are nowhere in the file. Someone who knows the village often
knows them at a glance. This path is therefore not a side feature but the main way the system
acquires data.

Contributions are applied **straight away** -- the immediate effect is what makes it appealing.
Three things catch the abuse case without slowing down the normal one:

  1. Only empty fields may be filled. What a curator set is untouchable.
  2. Coordinates must lie inside the region. Otherwise a photo ends up in the Pacific.
  3. Every change lands in ``changes`` and can be reverted individually in the admin area.
"""

import logging
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select, update
from sqlalchemy.orm import Session

from app.config import Settings, get_settings
from app.db import get_session
from app.models import Change, DatePrecision, Photo, PhotoStatus, Place, Source
from app.schemas import (
    DateContribution,
    HouseNumberContribution,
    LocationContribution,
    PhotoDetail,
    PlaceOut,
    TaskResponse,
)
from app.services import places
from app.services.dates import date_range, format_label
from app.services.needs import NEEDS, Need, open_filter
from app.services.places import ACCURACY_ADDRESS_M
from app.text import texts

log = logging.getLogger(__name__)
router = APIRouter(prefix="/contribute", tags=["contribute"])


@router.get("/next", response_model=TaskResponse, summary="A photo that is missing something")
def next_task(
    session: Annotated[Session, Depends(get_session)],
    need: Annotated[Need, Query(description="Which field should be missing")] = "location",
    exclude: Annotated[str, Query(description="Ids already shown, comma-separated")] = "",
    photo_id: Annotated[
        int | None, Query(description="Put up this photo, if it still owes this answer")
    ] = None,
) -> TaskResponse:
    """Return a random photo that is missing exactly this field.

    ``exclude`` holds the photos the visitor has just dismissed. Without that list the same image
    could reappear immediately, which feels broken.

    ``photo_id`` is a **wish, not an instruction**, and it comes from the detail view: whoever
    looks at one photograph full screen and taps "Wann war das?" means that one. Two rules keep the
    wish honest:

    * **It is checked against the same filter as any other photo.** A photo that no longer owes
      this answer -- because somebody else was quicker between the tap and this request -- would
      otherwise put a question on the screen that the write path then rejects with 409.
    * **Where it does not hold, the ordinary random pick runs.** Better a different photo than a
      dead end.

    ``exclude`` does *not* apply to it. Whoever asks for a photo by name may well have waved it
    away earlier and has now thought better of it.
    """
    skipped = {int(part) for part in exclude.split(",") if part.strip().isdigit()}

    filters = [Photo.status == PhotoStatus.PUBLISHED, open_filter(need)]
    open_count = session.scalar(select(func.count()).select_from(Photo).where(*filters)) or 0

    # What the *other* questions still hold: from it the screen decides whether "Weiss ich nicht"
    # still leads anywhere at all. With nothing else open the same photo would come back.
    #
    # A sum over all of them rather than the count of one -- with three questions a single count
    # would answer a narrower question than the panel is asking.
    open_other = sum(
        session.scalar(
            select(func.count())
            .select_from(Photo)
            .where(Photo.status == PhotoStatus.PUBLISHED, open_filter(other))
        )
        or 0
        for other in NEEDS
        if other != need
    )

    photo = None
    if photo_id is not None:
        photo = session.scalar(select(Photo).where(*filters, Photo.id == photo_id))

    if photo is None:
        query = select(Photo).where(*filters)
        if skipped:
            query = query.where(Photo.id.notin_(skipped))

        photo = session.scalar(query.order_by(func.random()).limit(1))

    # Everything seen: start over rather than reporting "nothing left", as long as anything is
    # still open at all.
    if photo is None and skipped and open_count:
        photo = session.scalar(select(Photo).where(*filters).order_by(func.random()).limit(1))

    return TaskResponse(
        need=need,
        open_count=open_count,
        open_other=open_other,
        photo=PhotoDetail.from_photo(photo) if photo else None,
    )


def _refinable(session: Session, photo: Photo) -> bool:
    """Is this exactly the photo the sharpening question is for?

    Asked of one photo rather than of the collection, so the same sentence decides who gets the
    question in the panel and who gets the buttons in the detail view. One rule, two readers.
    """
    return (
        session.scalar(
            select(func.count())
            .select_from(Photo)
            .where(Photo.id == photo.id, open_filter("housenumber"))
        )
        or 0
    ) > 0


@router.get(
    "/{photo_id}/housenumbers",
    response_model=list[PlaceOut],
    summary="House numbers a visitor may sharpen this photo to",
)
def photo_housenumbers(
    photo_id: int, session: Annotated[Session, Depends(get_session)]
) -> list[PlaceOut]:
    """The numbers of this photo's street -- **empty unless the photo may be sharpened at all**.

    That emptiness is the gate, and it is deliberately the only one: the detail view offers the
    picker when this list is not empty and needs no second rule of its own. A rule that lives in
    two places is a rule that will disagree with itself.
    """
    photo = _get_published_photo(session, photo_id)
    if not _refinable(session, photo):
        return []
    return [
        PlaceOut.from_place(place)
        for place in places.housenumbers_of(session, photo.place_name or "")
    ]


def _require_in_region(settings: Settings, lat: float, lon: float) -> None:
    """The pin can only be dropped on the map, and the map only shows the region -- check anyway.

    The API is reachable, and a photo in the Pacific would have vanished from the view without
    anyone noticing why. Pulled out rather than copied: the sharpening route takes its coordinate
    from the gazetteer, so it cannot really fail here -- but one rule with two callers stays one
    rule.
    """
    if (bbox := settings.region_bbox()) is None:
        return
    min_lon, min_lat, max_lon, max_lat = bbox
    if not (min_lat <= lat <= max_lat and min_lon <= lon <= max_lon):
        raise HTTPException(422, texts().contribute.outside_the_map)


def _require_empty(photo: Photo, field: str) -> None:
    """A visitor may only fill what is empty.

    Curated statements are untouchable -- and without this check the next visitor could overwrite
    the previous one's statement instead of both counting as confirmation.
    """
    taken = photo.lat is not None if field == "location" else photo.date_from is not None
    if taken:
        raise HTTPException(409, texts().contribute.already_stated)


def _get_published_photo(session: Session, photo_id: int) -> Photo:
    """A deleted photo answers 404 here, as in ``api/photos.py``.

    ``next_task`` never offers one, but the write routes take any id. A visitor's statement on a
    photo the curator took out would sit in the change log and wait for a moderator who never
    looks under „Gelöscht".
    """
    photo = session.get(Photo, photo_id)
    if photo is None or photo.status == PhotoStatus.DELETED:
        raise HTTPException(404, texts().photos.no_such_photo(photo_id))
    return photo


def _get_open_photo(session: Session, photo_id: int, field: str) -> Photo:
    photo = _get_published_photo(session, photo_id)
    _require_empty(photo, field)
    return photo


def _write_if(session: Session, photo: Photo, still: list, values: dict, refusal: str) -> None:
    """Write ``values`` only if the photo still is as the checks above found it.

    **The checks read the photo, and the write came a moment later.** Two visitors could both find
    the field empty -- at the kiosk and on the web instance, or through the API directly -- and the
    second one wrote over the first while both got a thank-you and both landed in the change log.
    That is the one thing ``_require_empty`` exists to prevent.

    So the condition travels into the UPDATE itself. SQLite writes one transaction at a time, and
    the second UPDATE sees what the first committed: it matches no row, and the route answers with
    the same 409 as if its check had caught it.
    """
    result = session.execute(
        update(Photo)
        .where(Photo.id == photo.id, Photo.status == PhotoStatus.PUBLISHED, *still)
        .values(**values)
        .execution_options(synchronize_session=False)
    )
    if result.rowcount == 0:
        raise HTTPException(409, refusal)


def _log_change(
    session: Session,
    photo: Photo,
    field: str,
    old: str | None,
    new: str,
    session_key: str | None,
    old_source: str | None = None,
) -> None:
    """Record the contribution for moderation.

    ``old`` is None for the two routes that only fill what is empty -- there is nothing to restore,
    and taking such a contribution back means clearing the field. The sharpening route is the one
    that overwrites, so it passes both the previous value and where it came from; without them a
    revert would drop the photo's location entirely and turn a curator's statement into a
    visitor's.
    """
    session.add(
        Change(
            photo_id=photo.id,
            field=field,
            old_value=old,
            old_source=old_source,
            new_value=new,
            source=Source.VISITOR,
            session_id=session_key,
        )
    )


@router.post("/{photo_id}/location", response_model=PhotoDetail, summary="Add a location")
def add_location(
    photo_id: int,
    contribution: LocationContribution,
    session: Annotated[Session, Depends(get_session)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> PhotoDetail:
    photo = _get_open_photo(session, photo_id, "location")
    _require_in_region(settings, contribution.lat, contribution.lon)

    values = {"lat": contribution.lat, "lon": contribution.lon, "location_source": Source.VISITOR}
    if contribution.place_name:
        values["place_name"] = contribution.place_name
    if contribution.accuracy_m is not None:
        values["location_accuracy_m"] = contribution.accuracy_m
    _write_if(session, photo, [Photo.lat.is_(None)], values, texts().contribute.already_stated)

    _log_change(
        session,
        photo,
        "location",
        None,
        f"{contribution.lat:.6f},{contribution.lon:.6f}"
        + (f" ({contribution.place_name})" if contribution.place_name else ""),
        contribution.session_id,
    )
    session.commit()
    session.refresh(photo)

    log.info("Visitor contribution: photo %s located", photo.id)
    return PhotoDetail.from_photo(photo)


@router.post(
    "/{photo_id}/housenumber",
    response_model=PhotoDetail,
    summary="Sharpen a street-precise photo to one house",
)
def add_housenumber(
    photo_id: int,
    contribution: HouseNumberContribution,
    session: Annotated[Session, Depends(get_session)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> PhotoDetail:
    """Move a photo from the middle of its street to one of its houses.

    **The exception to "visitors only fill what is empty"** (decisions.md, point 5) -- and it goes
    through its own door rather than loosening that check. ``_require_empty`` still reads exactly
    as it did; what stands beside it is a narrower rule: only street-precise, only to an address
    of that same street, and never the other way round.

    Curator statements may be sharpened too. That is a real widening, and it is why the change log
    carries the previous source: taking the contribution back has to give a curator's statement
    back to the curator.
    """
    photo = _get_published_photo(session, photo_id)
    if not _refinable(session, photo):
        raise HTTPException(409, texts().contribute.already_more_precise)

    address = session.get(Place, contribution.place_id)
    if address is None or address.kind != "adresse":
        raise HTTPException(404, texts().contribute.housenumber_unknown)
    if address.street != photo.place_name:
        raise HTTPException(422, texts().contribute.housenumber_wrong_street)

    _require_in_region(settings, address.lat, address.lon)

    street = photo.place_name
    previous_source = photo.location_source

    # Everything from the gazetteer row, nothing from the request. The condition is the statement
    # the checks above found, whole: the street, its accuracy, its coordinate and whose it was.
    # Anything less, and a second sharpening would log the street as the old value while it
    # replaced the first one's house.
    _write_if(
        session,
        photo,
        [
            Photo.place_name == street,
            Photo.location_accuracy_m.is_distinct_from(ACCURACY_ADDRESS_M),
            Photo.location_source.is_not_distinct_from(previous_source),
            Photo.lat.is_not_distinct_from(photo.lat),
            Photo.lon.is_not_distinct_from(photo.lon),
        ],
        {
            "lat": address.lat,
            "lon": address.lon,
            "place_name": address.name,
            "location_accuracy_m": ACCURACY_ADDRESS_M,
            "location_source": Source.VISITOR,
        },
        texts().contribute.already_more_precise,
    )

    _log_change(
        session,
        photo,
        "housenumber",
        street,
        address.name,
        contribution.session_id,
        old_source=previous_source,
    )
    session.commit()
    session.refresh(photo)

    log.info("Visitor contribution: photo %s sharpened to %s", photo.id, address.name)
    return PhotoDetail.from_photo(photo)


@router.post("/{photo_id}/date", response_model=PhotoDetail, summary="Add a year")
def add_date(
    photo_id: int,
    contribution: DateContribution,
    session: Annotated[Session, Depends(get_session)],
) -> PhotoDetail:
    photo = _get_open_photo(session, photo_id, "date")

    # The schema checks each part on its own -- a day from 1 to 31, a month from 1 to 12 -- and
    # not whether they make a date. 31 February passed it, and ``date()`` then raised a
    # ValueError that became a 500. The admin route has always caught it; this one had not.
    try:
        start, end, precision = date_range(
            contribution.year,
            contribution.month,
            contribution.day,
            DatePrecision(contribution.precision),
        )
    except ValueError:
        raise HTTPException(422, texts().contribute.no_such_date) from None
    _write_if(
        session,
        photo,
        [Photo.date_from.is_(None)],
        {
            "date_from": start,
            "date_to": end,
            "date_precision": precision,
            "date_source": Source.VISITOR,
        },
        texts().contribute.already_stated,
    )

    _log_change(
        session, photo, "date", None, format_label(start, end, precision), contribution.session_id
    )
    session.commit()
    session.refresh(photo)

    log.info("Visitor contribution: photo %s dated to %s", photo.id, photo.date_from)
    return PhotoDetail.from_photo(photo)

File: frontend/src/store/contribute.ts
/**
 * State of the "Hilf mit" panel.
 *
 * Kept apart from the map state because both run independently: the visitor can explore the map
 * while a question stands on the right, and the other way round.
 *
 * One exception -- locating a photo. There the pin is dropped on the very same map that shows the
 * photos. The pin therefore lives here and the map layer reads it.
 */

import { create } from "zustand";

import {
  NEEDS,
  type Need,
  type PhotoDetail,
  type Place,
  type Precision,
  type Task,
  fetchTask,
  postDate,
  postHouseNumber,
  postLocation,
} from "../api/client";
import { t } from "../text/de";
import { useKiosk } from "./kiosk";

/** How long the thank-you note stays before the next question arrives. */
const THANKS_MS = 2200;

/**
 * Session key.
 *
 * Distinguishes visitors at the same device without identifying them: a random value per page
 * load, stored nowhere. The curator can tell whether ten statements came from one person or from
 * ten -- and should not be able to do more than that.
 */
const SESSION_ID = Math.random().toString(36).slice(2, 12);

/**
 * Remember only the last few skipped photos, otherwise nothing would be left to show.
 *
 * One list for both questions, not one each: a photo somebody has just waved away should not come
 * straight back with the other question on it. That would read as "you were not listening".
 */
const SKIP_MEMORY = 20;

type ContributeState = {
  need: Need;
  task: Task | null;
  loading: boolean;
  error: string | null;
  /** Right after a contribution: thank-you note instead of the next question. */
  thanks: string | null;

  /** Photos the visitor has just dismissed. This session only. */
  skipped: number[];

  /**
   * Is a tap on the map allowed to drop the pin?
   *
   * Off by default: while "Wo ist das?" stands, somebody who only wants to look around -- pan the
   * map, get their bearings, find a photo nearby -- would otherwise answer the question by
   * accident, and one confirming tap later a place nobody meant is in the collection.
   *
   * **Lives here rather than in ``LocationTask``** so that it falls back on *every* route to the
   * next photo. The component is normally torn down on the way (``HelpPanel`` swaps it for
   * ``DateTask``), which would clear a ``useState`` by itself -- but not when ``load`` falls back
   * to the question it started from because the other one has run dry. That case arrives exactly
   * when one kind of gap has been worked off, and it would leave the map armed over a photo the
   * visitor has not looked at yet.
   */
  pickingOnMap: boolean;

  /** Pin dropped on the map while the location question is running. */
  pin: { lat: number; lon: number } | null;
  /** Name from the place search, when set that way. */
  pinLabel: string | null;
  /**
   * How precise the pin is, in metres -- a house number is worth more than a street.
   *
   * Null for a pin someone tapped onto the map: how well they aimed is not for us to claim.
   * Dragging the pin clears it again, for the same reason.
   */
  pinAccuracy: number | null;

  /**
   * The house numbers currently on the buttons -- the map draws them beside the photos.
   *
   * Lives here rather than in the component for the same reason the pin does: the map is a
   * different corner of the tree, and it has to read what the panel is offering.
   */
  offeredNumbers: Place[];
  setOfferedNumbers: (numbers: Place[]) => void;

  load: (need?: Need) => Promise<void>;
  /** Put one named photo up for one question -- the way out of the detail view. */
  askAbout: (photoId: number, need: Need) => Promise<void>;
  skip: () => void;
  setPickingOnMap: (on: boolean) => void;
  setPin: (
    pin: { lat: number; lon: number } | null,
    details?: { label?: string | null; accuracyM?: number | null },
  ) => void;
  submitLocation: () => Promise<void>;
  submitDate: (year: number, precision: Precision) => Promise<void>;
  submitHouseNumber: (placeId: number) => Promise<void>;
};

let abort: AbortController | null = null;
let thanksTimer: ReturnType<typeof setTimeout> | null = null;

/**
 * "housenumber" sharpens "location" -- and therefore never follows it.
 *
 * The one real exception to the plain ranking, and it is worth a line of its own: somebody who
 * has just pressed "Reicht so — die Straße genügt" must not be asked "Genauer: welche
 * Hausnummer?" in the same breath. That reads as though nobody had been listening.
 */
const refines: Partial<Record<Need, Need>> = { housenumber: "location" };

/**
 * Where to go from here: every other question, in rank order.
 *
 * The ranking *is* the array -- "sharpening comes last" needs no case distinction. Callers walk
 * the list and take the first that leads anywhere.
 */
export function nextNeeds(current: Need): Need[] {
  return NEEDS.filter((need) => need !== current);
}

/**
 * Where to go after somebody has **answered** ``current`` -- the same list, minus the question
 * their answer already settled.
 *
 * The distinction matters and is easy to lose: running dry and being answered are not the same
 * thing. ``nextNeeds`` covers a question that had nothing left, where every other one is fair
 * game. Only an answer can make a further question redundant.
 */
export function nextAfterAnswer(current: Need): Need[] {
  return nextNeeds(current).filter((need) => refines[need] !== current);
}

/**
 * Does this photo still owe an answer to that question?
 *
 * **Not for "housenumber", and that is not an oversight.** Whether a photo can be sharpened
 * depends on the gazetteer -- whether its street holds addresses at all -- and a photo object has
 * never seen it (see `services/needs.py`). The chain, which hands the same photo on to the next
 * question, therefore never leads into the sharpening question; the panel reaches it through a
 * fresh task instead.
 */
export function stillNeeds(photo: PhotoDetail, need: Need): boolean {
  if (need === "location") return photo.needs_location;
  if (need === "date") return photo.needs_date;
  return false;
}

/**
 * The thank-you for an answered question.
 *
 * Only where the photo is complete may it name what became visible. Where the same photo carries
 * on, the sentence asks the question that really follows -- a promise nobody can see kept is
 * worse than no sentence at all.
 */
export function thanksFor(answered: Need, chained: Need | null): string {
  return chained ? t.help.thanksAsk[chained] : t.help.thanks[answered];
}

export const useContribute = create<ContributeState>((set, get) => {
  /**
   * Put up the next task, trying the questions in the order given.
   *
   * **``order`` rather than a single question**, because falling back is what makes taking turns
   * safe: the kinds empty out at different rates, and in a collection where every photo is placed
   * but half of them are undated, always asking "where is this?" would report "everything is
   * complete" while hundreds of photos still wait for a year.
   *
   * That the caller hands over the whole order is the point. Only the caller knows *why* it is
   * loading -- and a question that was just answered has to stay out of the fallback too, not only
   * out of the first attempt (see ``nextAfterAnswer``). With a single question plus a fallback
   * computed here, "Reicht so — die Straße genügt" was answered with "Welche Hausnummer?" two
   * seconds later.
   *
   * ``prefer`` carries a photo that has just been contributed to. If it still owes an answer to
   * the first question, it is put up instead of the random one -- see the chain in
   * ``contribute()``.
   *
   * ``wanted`` is the other direction: an id somebody asked for, out of the detail view. It goes
   * to the server, which checks it and falls back where it no longer holds -- **only for the first
   * question**, because the wish was for that one. Where the fallback questions run, it is a
   * wish about a photo that has nothing left to say on them.
   */
  async function load(order: Need[], prefer?: PhotoDetail | null, wanted?: number | null) {
    abort?.abort();
    abort = new AbortController();
    const signal = abort.signal;

    // Changing photo or question puts the map back where the visitor had it. Covers "Weiss ich
    // nicht" as well; after a contribution the thank-you timer has already done it, and doing it
    // twice does no harm.
    useKiosk.getState().releaseFocus();

    set({
      loading: true,
      error: null,
      pin: null,
      pinLabel: null,
      pinAccuracy: null,
      pickingOnMap: false,
      offeredNumbers: [],
    });
    try {
      const need = order[0]!;
      const task = await fetchTask(need, get().skipped, wanted, signal);

      // The counts come from the fetch, only the photo is swapped -- so the panel keeps saying
      // truthfully how much is still open. Deliberately ignores ``skipped``: whoever waved a
      // photo away earlier and has now told us something about it after all gets it back with
      // the other question, and "Weiß ich nicht" remains the way out.
      if (prefer && stillNeeds(prefer, need)) {
        set({ task: { ...task, photo: prefer }, need, loading: false });
        return;
      }

      if (!task.photo) {
        for (const other of order.slice(1)) {
          const fallback = await fetchTask(other, get().skipped, null, signal);
          if (fallback.photo) {
            set({ task: fallback, need: other, loading: false });
            return;
          }
        }
      }

      set({ task, need, loading: false });
    } catch (e) {
      if (signal.aborted) return;
      set({ loading: false, error: e instanceof Error ? e.message : String(e) });
    }
  }

  function showThanks(text: string, order: Need[], prefer: PhotoDetail | null) {
    if (thanksTimer) clearTimeout(thanksTimer);
    set({ thanks: text });
    thanksTimer = setTimeout(() => {
      thanksTimer = null;
      set({ thanks: null });
      // Map and time range come back as the thank-you goes -- both live exactly as long as it
      // does, without a second timer.
      useKiosk.getState().releaseFocus();
      void load(order, prefer);
    }, THANKS_MS);
  }

  /**
   * Apply a contribution, then thank -- and hand the same photo on where something is still open.
   *
   * **The chain is what keeps the thank-you honest.** "Das Foto ist jetzt auf der Zeitleiste" is a
   * promise the view cannot keep for a photo without a place: ``showPhoto()`` leaves map and
   * slider alone, so the visitor reads a sentence and sees nothing. Where something is missing,
   * the thank-you therefore asks about it instead of claiming anything -- and the next question is
   * about *this* photo rather than a random one. Somebody who has just shown they know this
   * picture, and is looking straight at it, is the best moment the panel ever gets.
   *
   * It ends by itself: once nothing is missing, ``stillNeeds`` is false and the next photo comes.
   */
  async function contribute(action: (photo: PhotoDetail) => Promise<PhotoDetail>) {
    const { task, need } = get();
    if (!task?.photo) return;

    set({ loading: true, error: null });
    try {
      // The backend hands back the updated photo -- place and date therefore come first hand
      // instead of being guessed at again here.
      const updated = await action(task.photo);
      set({ loading: false, pin: null, pinLabel: null, pinAccuracy: null, pickingOnMap: false });

      // Map and timeline have to show it now. The thank-you note promises exactly that -- and
      // before this line it only came true once somebody happened to pan the map.
      useKiosk.getState().refresh();
      // And the view settles on this one photo for as long as the thank-you stands.
      useKiosk.getState().showPhoto(updated);

      // The first of the remaining questions this very photo can still answer comes first; the
      // rest follow as the fallback, with the answered question last of all. What is *not* in the
      // list is the question this answer settled -- see ``nextAfterAnswer``.
      const others = nextAfterAnswer(need);
      const chained = others.find((other) => stillNeeds(updated, other)) ?? null;
      const order = chained ? [chained, ...others.filter((o) => o !== chained)] : others;
      showThanks(thanksFor(need, chained), [...order, need], chained ? updated : null);
    } catch (e) {
      // Most common case: somebody else was quicker (HTTP 409). The backend already phrases that
      // message kindly.
      set({ loading: false, error: e instanceof Error ? e.message : String(e) });
    }
  }

  return {
    // The highest-ranked question, not a name written out again -- the ranking is the array.
    need: NEEDS[0],
    task: null,
    loading: false,
    error: null,
    thanks: null,
    skipped: [],
    pickingOnMap: false,
    pin: null,
    pinLabel: null,
    pinAccuracy: null,
    offeredNumbers: [],

    setOfferedNumbers(numbers) {
      set({ offeredNumbers: numbers });
    },

    load(need) {
      const first = need ?? get().need;
      return load([first, ...nextNeeds(first)]);
    },

    /**
     * "Wann war das?" on the photo somebody is looking at.
     *
     * The detail view no longer answers anything itself; it branches here (see decisions.md). What
     * arrives is an ordinary task, and **everything after it is the ordinary flow** -- thank-you,
     * the chain to whatever this photo still lacks, then a new photo. That there is no special
     * case for the return trip is the design, not an omission: whoever answers out of a photo has
     * landed in the panel, and that is where the next question belongs.
     *
     * The ranking stays the fallback. Should the photo no longer owe this answer -- somebody else
     * was quicker -- the server hands over a different one, and the panel carries on rather than
     * standing there with a question that has already been settled.
     */
    askAbout(photoId, need) {
      return load([need, ...nextNeeds(need)], null, photoId);
    },

    /**
     * "Weiß ich nicht -- nächstes Foto" changes the question, not just the picture.
     *
     * Whoever cannot place a photo may well know the decade, and the other way round. Asking the
     * same kind of question again is what makes a visitor give up after three pictures. Where the
     * next question has nothing left, ``load`` carries on down the ranking.
     */
    skip() {
      const { task, need, skipped } = get();
      const id = task?.photo?.id;
      set({
        skipped: id ? [...skipped, id].slice(-SKIP_MEMORY) : skipped,
        pin: null,
        pinLabel: null,
        pinAccuracy: null,
        pickingOnMap: false,
      });
      // The skipped question last: it stays reachable with a different photo, but only once
      // everything else has run dry.
      void load([...nextNeeds(need), need]);
    },

    setPickingOnMap(on) {
      set({ pickingOnMap: on });
    },

    setPin(pin, details) {
      set({
        pin,
        pinLabel: details?.label ?? null,
        pinAccuracy: details?.accuracyM ?? null,
      });

      // Out of the place search -- the only thing that sets a label -- the map takes the visitor
      // along: they did not place the pin themselves and want to see where it landed. A pin
      // tapped onto the map or dragged leaves it alone: that is where they just aimed, and a map
      // that jumps out from under a finger feels like slipping.
      if (pin && details?.label) useKiosk.getState().showLocation(pin.lat, pin.lon);
      // "Punkt entfernen" takes the zoom back too.
      if (!pin) useKiosk.getState().releaseFocus();
    },

    async submitLocation() {
      const { pin, pinLabel, pinAccuracy } = get();
      if (!pin) return;
      await contribute((photo) =>
        postLocation(photo.id, {
          lat: pin.lat,
          lon: pin.lon,
          ...(pinLabel ? { place_name: pinLabel } : {}),
          ...(pinAccuracy !== null ? { accuracy_m: pinAccuracy } : {}),
          session_id: SESSION_ID,
        }),
      );
    },

    async submitDate(year, precision) {
      await contribute((photo) => postDate(photo.id, { year, precision, session_id: SESSION_ID }));
    },

    async submitHouseNumber(placeId) {
      await contribute((photo) =>
        postHouseNumber(photo.id, { place_id: placeId, session_id: SESSION_ID }),
      );
    },
  };
});
